SixTrack/tools/astuce/astuce++.cpp

```diff
@@ -0,0 +1,708 @@
+		std::string FlagName = FixCase(line->Text.substr(NameStart,NameEnd-NameStart));
+
+		size_t Position = 0;
+		size_t NextOR = 0;
+		size_t NextAND = 0;
+		std::vector<size_t> Splits;
+		std::vector<std::string> FlagNames;
+		std::vector<bool> And;
+		std::vector<bool> Or;
+		std::vector<bool> States;
+
+		Splits.push_back(Position);
+		while(NextOR != std::string::npos && NextAND != std::string::npos)
+		{
+			NextOR = FlagName.find(".OR.", Position);
+			NextAND = FlagName.find(".AND.", Position);
+			Position = std::min(NextOR, NextAND);
+			if(Position != std::string::npos)
+			{
+				Splits.push_back(Position);
+			}
+			Position++;
+		}
+		Splits.push_back(FlagName.size());
+
+		for(size_t n=1; n < Splits.size(); n++)
+		{
+			std::string SplitSubString = FlagName.substr(Splits.at(n-1),Splits.at(n));
+			FlagNames.push_back(SplitSubString);
+		}
+
+		for(size_t n=0; n < FlagNames.size(); n++)
+		{
+			bool NOT = false;
+			bool AND = false;
+			bool OR = false;
+
+			if(FlagNames.at(n).substr(0,5) == ".AND.")
+			{
+				AND = true;
+				FlagNames.at(n).erase(0,5);
+			}
+			if(FlagNames.at(n).substr(0,4) == ".OR.")
+			{
+				OR = true;
+				FlagNames.at(n).erase(0,4);
+			}
+			if(FlagNames.at(n).substr(0,5) == ".NOT.")
+			{
+				FlagNames.at(n).erase(0,5);
+				NOT = true;
+			}
+
+			//Enable/disable lines
+			std::set<std::string>::const_iterator Flags_itr = Flags.find(FlagNames.at(n));
+			if(Flags_itr != Flags.end())
+			{
+				//flag is found -> enabled
+				if(NOT == false){EnabledState = true;}
+				else{EnabledState = false;}
+			}
+			else
+			{
+				//flag is NOT found -> enabled
+				//.not.xxx and xxx is not enabled
+				if(NOT == false){EnabledState = false;}
+				else{EnabledState = true;}
+			}
+			Or.push_back(OR);
+			And.push_back(AND);
+			States.push_back(EnabledState);
+		}
+
+		//Do the first
+		EnabledState = States.at(0);
+
+		for(size_t n = 1; n < States.size(); n++)
+		{
+			if(And.at(n) == true)
+			{
+				EnabledState = (EnabledState & States.at(n));
+			}
+			else if(Or.at(n) == true)
+			{
+				EnabledState = (EnabledState | States.at(n));
+			}
+			else
+			{
+				//Should never hit this
+				std::cerr << "ERROR: in .if. .and. .or. processing - Multiple entries and neither .and. or .or. true at entry " << n << std::endl;
+				exit(EXIT_FAILURE);
+			}
+		}
+
+		//AND (&) the result of the above with the current stack top
+		EnabledStates.push(EnabledState & EnabledStates.top());
+	}
+	else if(line->Text.substr(0,3) == "+ei")
+	{
+		//Disable this line
+		line->Enabled = false;
+
+		EnabledStates.pop();
+	}
+	else
+	{
+		//If a general line is to be enabled or disabled, this depends on the current state
+		line->Enabled = EnabledStates.top();
+		return;
+	}
+}
+
+void Astuce::ExpandCA()
+{
+	bool Replaced = true;
+	while(Replaced)
+	{
+		//Repeat until no +ca exists
+		Replaced = false;
+
+		//First step, go through existing +cd blocks and check for +ca
```

Comment:
This would be a description of the whole while loop? Replacement inside decks is below.